File: CS50X/Week6/credit.py
import re
import sys


def main():
    while True:
        number = input("Number: ").strip()
        try:
            int(number)
            break
        except:
            continue
    is_valid = validate(number)
    if is_valid == False:
        sys.exit("INVALID")
    else:
        print(is_valid)


def validate(s):
    if len(s) < 13 or len(s) > 16:
        return False
    elif len(s) == 14:
        return False
    # check if checksome ends with 0
    check = checksum(s)
    if check == False:
        sys.exit("INVALID")
    # check if pattern matches any card issuer
    if matches := re.match(r"^(4|5[1-5]|3[4|7])", s):
        if matches.group(1)[0] == "4" and (len(s) == 13 or len(s) == 16):
            return "VISA"
        elif matches.group(1)[0] == "3" and len(s) == 15:
            return "AMEX"
        elif matches.group(1)[0] == "5" and len(s) == 16:
            return "MASTERCARD"
        else:
            return False
    else:
        return False


def checksum(number):
    i = 0
    sum_first = 0
    sum_second = 0
    # reverse card number
    number = number[::-1]

    for digit in number:
        digit = int(digit)

        # every second from first
        if i % 2 == 1:
            if digit * 2 > 9:
                sum_first += int(str(digit * 2)[0])
                sum_first += int(str(digit * 2)[1])
            else:
                sum_first += digit * 2
        # every second after second
        else:
            sum_second += digit
        i += 1

    total = sum_first + sum_second
    if total % 10 == 0:
        return True
    else:
        return False


if __name__ == "__main__":
    main()

File: CS50X/Week0-1_C/credit/credit.c
#include <cs50.h>
#include <stdio.h>
int main()
{
    long num;
    int counter = 0;
    counter = 0;
    num = get_long("Number: ");
    long holder = num;
    // Counter to not accept too long or too short numbers
    while (holder > 0)
    {
        holder /= 10;
        counter++;
    }

    if (counter < 13 || counter > 17)
    {
        printf("INVALID\n");
    }
    else
    {
        // Save last figit of number
        int tot = num % 10;
        // Used to ge every second from first.
        long num_for_firsts = num / 10;
        // Used in llop to get first digit
        long first_digits = num;
        int sum = 0;

        // Loop to get every second number starting from second digit.
        while (num != 0)
        {
            num = num / 10;
            int digit = num % 10;
            num = num / 10;
            int multy = digit * 2;
            // Loop to get digits if after multiplication exceeded 10.
            while (multy != 0)
            {
                int one = multy % 10;
                multy = multy / 10;
                sum += one;
            }
        }

        // Loop to get every second digit statrting from first digit.
        while (num_for_firsts != 0)
        {
            num_for_firsts = num_for_firsts / 10;
            int first = num_for_firsts % 10;
            num_for_firsts = num_for_firsts / 10;

            tot += first;
        }

        sum += tot;

        // secont digit off summuary to check validity.
        int second_of_sum = sum % 10;
        int digits;

        // Loop which gets only first two digits of number to recognize card issuer.
        while (first_digits > 10)
        {
            int dig = first_digits % 100;
            first_digits = first_digits / 10;
            digits = dig;
        }

        if (second_of_sum == 0)
        {
            if (counter == 15)
            {
                if (digits == 34 || digits == 37)
                {
                    printf("AMEX\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            else if (counter == 16)
            {
                if (digits == 51 || digits == 52 || digits == 53 || digits == 54 || digits == 55)
                {
                    printf("MASTERCARD\n");
                }
                else if (digits / 10 == 4)
                {
                    printf("VISA\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            else if (counter == 13)
            {
                if (digits / 10 == 4)
                {
                    printf("VISA\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
}
